### ISSUE TITLE
[rulename] issue title

### ISSUE THREAD

**Issue by xenoterracide** [typescript-eslint/typescript-eslint#4908]
I'm probably just missing something but seems odd given the docs, I also tried setting `"allowNullableObject": true` manually
```
  33:10  warning  Unexpected object value in conditional. The condition is always true                  @typescript-eslint/strict-boolean-expressions
```

**Repro**
```sh
npm exec -- eslint --debug '**/src/**/*.ts'
```

the plugin is saying that orderByClause is the unexpected object
```ts
    private readonly orderByClause?: SpecFactory<T, OrderByClause<T>>
    const simple = this.orderByClause ? [...specs, this.orderByClause()] : specs
```

```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "strict": true,
    "outDir": "../test-dist"
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "dist", "test-dist"]
}
```
```json
{
  "compilerOptions": {
    "strict": false,
    "resolveJsonModule": true,
    "module": "commonjs",
    "esModuleInterop": true,
    "target": "es6",
    "noImplicitAny": true,
    "moduleResolution": "node",
    "sourceMap": true,
    "outDir": "../dist",
    "baseUrl": "./",
    "lib": ["ES2018"],
    "types": ["mocha"],
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "paths": {
      "*": ["node_modules/*", "*"]
    },
    "plugins": [
      {
        "name": "typescript-strict-plugin"
      }
    ]
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "**/*.spec.ts", "dist"]
}
```

```JSON
{
  "root": true,
  "env": {
    "es6": true,
    "mocha": true
  },
  "extends": [
    "eslint:recommended",
    "plugin:lodash-fp/recommended",
    "plugin:import/recommended",
    "plugin:import/typescript",
    "plugin:@typescript-eslint/recommended",
    "plugin:editorconfig/all",
    "prettier"
  ],
  "globals": {
    "Atomics": "readonly",
    "SharedArrayBuffer": "readonly"
  },
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module",
    "project": ["*/tsconfig.eslint.json", "packages/*/tsconfig.eslint.json"]
  },
  "plugins": ["@typescript-eslint", "lodash-fp", "import", "simple-import-sort", "editorconfig", "prettier"],
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolvers": {
      "typescript": {
        "project": ["*/tsconfig.json", "packages/*/tsconfig.json"]
      }
    }
  },
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": ["warn", { "allowString": false, "allowNumber": false }],
    "lodash-fp/use-fp": "warn",
    "import/no-unresolved": [
      "error",
      {
        "ignore": ["inversify-express-utils/dts/results"]
      }
    ],
    "no-unused-vars": "off",
    "simple-import-sort/imports": "error",
    "simple-import-sort/exports": "error",
    "import/first": "error",
    "import/newline-after-import": "error",
    "import/no-duplicates": "error",
    "no-restricted-imports": [
      "error",
      {
        "patterns": [
          {
            "group": ["*/dist/*"]
          }
        ]
      }
    ],
    "@typescript-eslint/explicit-function-return-type": [
      "warn",
      {
        "allowExpressions": true,
        "allowTypedFunctionExpressions": true
      }
    ],
    "@typescript-eslint/no-unused-vars": [
      "error",
      {
        "argsIgnorePattern": "^_"
      }
    ],
    "prettier/prettier": ["error", { "requirePragma": true }]
  },
  "overrides": [
    {
      "files": ["*.spec.ts"],
      "rules": {
        "no-unused-expressions": "off",
        "@typescript-eslint/no-empty-function": "off"
      }
    }
  ]
}
```

**Expected Result**

There shouldn't be a warning

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.19.0` |
| `@typescript-eslint/parser`        | `5.19.0` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `v16.14.2` |

**Comment by JoshuaKGoldberg** [comment]
Hi @xenoterracide, thanks for filing an issue!

The code snippet you posted isn't enough information for us to investigate the rule. We don't know what `SpecFactory` or `OrderByClause` are meant to look like, nor what the rest of the class is. Please give us a reproduction we can use to help you. 🙂

**Comment by xenoterracide** [comment]
updated the example, was doing that anyways ;)

**Comment by JoshuaKGoldberg** [comment]
I don't get the lint error with that snippet 😢. Please provide a **full isolated** reproduction such as a https://typescript-eslint.io/play link or as a GitHub repo with a clear set of instructions.

The closest I could get was this: https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=GYVwdgxgLglg9mABAIxDANgEwBQAcCGA5gKYBKxAjiMQM5QD8AXInMgFbHQCUzA3ouhgBbGFGZgQQ5MQBOiAL6JeAWABQiDYhnEoIGUhXrNxwSLGICJclVpRE9RAEZEzcJmLAYYYpgA0a43k1INUgA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AhOktgEYD2NucAhgTgB4wqIJo0EIgAuANog4KFDRQgANKEa5cNAO4A5AK4LGVBgHkqAKziVB5NXAC+AXXNA&tsConfig=N4IgzgLgTglgxhEAuaBXApgXyA -- note the ESLint config & TypeScript configs in it.

**Comment by bradzacher** [comment]
The eslint version should not influence the output of our lint rules at all.

Please create an isolated repro repo to help us investigate this bug for you. It's possible you have a complex interaction between types.

**Comment by xenoterracide** [comment]
something sure, but I think that because an `npm upgrade eslint` made the warn go away, but it also updated other things... also, now I'm very confused...

```
❯ npm explain eslint | grep -v '^ '
eslint@7.32.0 dev
node_modules/eslint

eslint@8.14.0 dev
node_modules/eslint-plugin-editorconfig/node_modules/eslint
❯ npm exec eslint --version
8.6.0
```

**Comment by JoshuaKGoldberg** [comment]
Well, looks like it's not a problem on our end. Good look with your versioning shenanigans! 😄 

Please do post back if you do identify something wrong with typescript-eslint.

**Comment by xenoterracide** [comment]
yep, sorry about that... I'm suggesting we move to yarn pnmp, because whatever the node module resolution is doing is nuts
